DON'T MERGE YET: Content for CML workflow in Studio (#4034)
iterative/dvc.org: content/docs/studio/user-guide/projects-and-experiments/run-experiments.md

Iterative Studio uses your regular CI/CD setup (e.g. GitHub Actions) to run the
experiments. This means that to enable experimentation from Iterative Studio,
you need to do the following:

1. First, integrate your Git repository with a CI/CD setup that includes model

>    training process. You can do this in one of two ways:

Reviewer: model training process -> job execution.

Author: This has been changed to link to the prepare-repositories page. PTAL.

   - by
     [using the wizard provided by Iterative Studio](#use-the-iterative-studio-wizard-to-set-up-your-ci-action).
   - by creating the CI set up on your own. For this, create workflow files
     (such as GitHub Actions) that get triggered on push or pull request. For
     more details on how to set up
     [CI/CD pipelines for your ML project](/doc/use-cases/ci-cd-for-machine-learning),

>      refer to [CML](https://cml.dev).

Reviewer 2: This (#1) is pretty much a repeat of [Prepare Your Repositories to Run New Experiments](https://dvc-org-cml-in-studio-bgeulj3d.herokuapp.com/doc/studio/user-guide/prepare-your-repositories#prepare-your-repositories-to-run-new-experiments). Why not just link instead?

Author: Done.

2. Then,
   [submit your experiment from Iterative Studio](#submit-a-new-experiment). The
   submission will invoke your CI/CD setup, triggering the model training
   process.

> Due to access restrictions, you cannot run experiments on the demo project
> (`example-get-started`) that is provided to you by default. Once you connect
> to your ML project repositories, you can follow the instructions given below
> to run experiments directly from Iterative Studio.

## Use the Iterative Studio wizard to set up your CI action

> When you select one of the commits in your project table and click on the `Run`
> button, an input form will open. At the top of this form, you will see a small
> message that invites you to set up your CI in case you have not done it yet.

Reviewer 2: I feel like we tend to be really verbose in Studio docs. This for example could be summarize quite a bit, even to this extreme:

```suggestion
Select a commit and click **Run**, the CI setup will start if needed.
```

p.s. Shortening instructions like that means many of the paragraphs (turn short sentences) could be collapsed into a single paragraph.

Author: Updated.

Reviewer 2: I like it 😬

> ![](https://static.iterative.ai/img/studio/set_up_cml_message.png)

Reviewer 2: It's not being rendered, not sure why. See https://dvc-org-cml-in-studio-bgeulj3d.herokuapp.com/doc/studio/user-guide/projects-and-experiments/run-experiments#use-the-iterative-studio-wizard-to-set-up-your-ci-action

May need to check with **websites** whether this will be a problem on prod.

Author: I see it now. Can you pls recheck?

Reviewer 2: I'm seeing the images now 🙂

> Click on `Click here` to use the wizard to create your CML workflow within your
> CI/CD vendor.
> 
> The wizard UI is divided in two sections:

Reviewer 2: Too verbose?

```suggestion
The CML wizard is divided in two sections:
```

Reviewer 2: > Click on `Click here`

It feels like that should be "self-documenting" -- we don't need to explain to users that "click here" means they should click there.

Bottom line: Studio is a visual product so it should be intuitive, then we don't need to document every step in such detail because it just repeats the UI.

Author: Updated. And I'll see what else can be shortened.

- Left section with 2 sets of parameters:

  1. [Configuration of your self-hosted runner, which is used in the `Deploy runner` step of your CI process](#step-1-deploy-runner)

>   2. [Model training script, which is used in the `Train` step of your CI process](#step-2-train)

Author: Updated. PTAL.

> - Right section which displays in real-time the

Reviewer: `in real time` sounds too loud for client-side yaml generation 🙃

Reviewer 3: But it is! we had to do the jsons 😁

Author: Have removed `real time`. Although it is real time (even if it may not be a huge deal 🙂 ).

> The wizard UI is divided in two sections:
> 
> - Left section with 2 sets of parameters:
> 
>   1. [Configuration of your self-hosted runner, which is used in the `Deploy runner` step of your CI process](#step-1-deploy-runner)
>   2. [Model training script, which is used in the `Train` step of your CI process](#step-2-train)
> 
> - Right section which displays in real-time the
>   [generated yaml to be used in your CI set up](#ci-yaml).
> 
> By default, some standard values are assumed for all the inputs and a complete
> workflow (based on these default values) is ready for you to copy.

Reviewer 2: Another summarizing example:

```suggestion
The CML wizard is divided in two sections prefilled with default values:
[configuring your self-hosted runner] and [model training script]. The right panel will display
the [generated CML config] in real-time.
```

Author: Changed, but different from your suggestion. PTAL.

![](https://static.iterative.ai/img/studio/set_up_cml_full.png)

### Step 1: Deploy runner

This step is responsible for launching a self-hosted runner within your cloud
vendor. The parameters listed here are a subset of the parameters for
[CML self-hosted runners](https://cml.dev/doc/self-hosted-runners).

- `Spot`: Whether you want to launch a spot cloud instance, cutting down the
  costs of your training.

> - `Reuse`: Values for the CML flags `reuse` and `reuse-idle`. See all

Reviewer 2: Consider making a table.

Author: Done.

  [CML options](https://cml.dev/doc/ref/runner#options) for details.

- `Labels`: Text labels to identify your CML runners from other self hosted
  runners that you might have.

- `Cloud`: Your cloud provider.

- `Region`: Cloud-vendor specific region or a CML synthetic region (an
  abstraction across all the cloud vendors).

- `Type`: Cloud-vendor specific instance type or a CML synthetic type
  `M`/`L`/`XL` (an abstraction across all the cloud vendors).

  `Type` is also tied to GPU behavior. If you choose an instance with a
  selectable GPU (such as a CML instance type or any GCP instance), the `GPU`
  parameter will show up.

- `HDD size`: Hard disk size in GB. We highly recommend you to enter a big
  enough value (eg, 100) to avoid unexpected runner termination due to hard disk
  exhaustion.

### Step 2: Train

> This is the script needed for your runner to train your model. The default

Reviewer 2: Use same format as Step 1 (either bullet list or table) ? Even if it's just one thing.

Author: Done.

template is a very common combination of CML and DVC taking into account that
DVC enables you to make the most of Iterative Studio. You can update this script
to reflect your exact model training process, whether you use DVC or not.

### Generated CI yaml

> The right section displays the generated CI yaml content, which reflects all

Reviewer 2: This whole section could be skipped IMO. We already mentioned it and said it's for copying. Describing every button is unnecessary (Studio is a visual product -- should be self-documenting I think).

Author: Done. PTAL.

your input parameters.

Above the yaml textarea, you will find a `Copy to clipboard` link. Once you have
specified all your parameters, click on this link to copy the yaml content.

Then, click on `paste in your CI Workflow file`. This will open the editor in

> your Git provider. Paste the content here to create your CI script.

Author: This entire section has now been removed. PTAL.

> That's it! At this point you should have CML in place within your CI/CD to run
> your experiments. After this, proceed with submitting your experiments as
> described below.

Reviewer 3: ```suggestion
There is yet a final and crucial step. In other to be able to work effectively with your repo and be able to launch the runner in your desired cloud provider: Setup the yaml workflow ENV vars as secrets. You can learn how to do this simple step [here](https://cml.dev/doc/self-hosted-runners#environment-variables).

That's it! At this point you should have CML in place within your CI/CD to run
your experiments. After this, proceed with submitting your experiments as
described below.
```

> described below.

Reviewer 2: Well, this info (summarized) can probably stay somewhere. It probably fits in the H2's intro (before Step 1).

Author: Have removed this section, and added a bullet point (# 2) in the `How Iterative Studio runs ML experiments` section on top. PTAL. 
cc @DavidGOrtega pls review this change.
